design/external-processing-design.md

```diff
@@ -0,0 +1,619 @@
+
+
+## High-Level Design
+
+Contour will add HTTP support for Envoy's External Processing.
+
+
+new type: `ExternalProcessor` and its friends: `ExtProc`, `ExtProcOverride`, `ProcessingMode`, `HeaderMutationRules`,`GRPCService`, `ProcessingPhase`,`ExtProcPolicy`, will be defined for implement the design.
+
+A `globalExtProc` config would define a global external processing configuration for all hosts and routes.
+
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ContourConfiguration
+...
+spec:
+  globalExtProc:
+    processors:
+      - grpcService:
+          extensionRef:
+            apiVersion: projectcontour.io/v1alpha1
+            name: extproc-extsvc
+            namespace: extproc-test
+          failOpen: false
+          responseTimeout: 30s
+        processingMode:
+          requestBodyMode: 0
+          requestHeaderMode: 1
+          requestTrailerMode: 2
+          responseBodyMode: 0
+          responseHeaderMode: 1
+          responseTrailerMode: 2
+      - grpcService:
+          extensionRef:
+            apiVersion: projectcontour.io/v1alpha1
+            name: extproc-extsvc2
+            namespace: extproc-test
+          failOpen: true
+          responseTimeout: 60s
+        processingMode:
+          requestBodyMode: 0
+          requestHeaderMode: 1
+          requestTrailerMode: 2
+          responseBodyMode: 0
+          responseHeaderMode: 1
+          responseTrailerMode: 2
+...
+```
+### Opting out from Global/VirtualHost External Processing
+
+By default, enabling `Global` external processing would allow custom HTTP message processing on all virtual hosts. On the other hand, if processing is done at the `virtualHost` level, custom HTTP message processing would only be enabled for the specific virtual host in question. However, individual owners of HTTPProxy will have the flexibility to modify or override this setting.
+
+#### Override/Disabling External Processing
+
+##### Virtual Host level
+
+`Override(Only for HTTPS)/Disable` the global external processing on the virtual host. This setting would **override/disable** all routes on said virtual host.
+
+```yaml
+kind: HTTPProxy
+...
+spec:
+  virtualhost:
+    extProc:
+      extProcPolicy:
+        disabled: false  # true: for disable
+      processors:
+        - grpcService:
+            extensionRef:
+              apiVersion: projectcontour.io/v1alpha1
+              name: extproc-extsvc3
+              namespace: extproc-test
+            failOpen: true
+            responseTimeout: 60s
+          processingMode:
+            requestBodyMode: 0
+            requestHeaderMode: 1
+            requestTrailerMode: 2
+            responseBodyMode: 0
+            responseHeaderMode: 1
+            responseTrailerMode: 2
+...
+```
+
+##### Route level
+
+For more precise control, the `Global` and/or `VirtualHost` external processing can also be **overrideed/toggled** on an individual route. the priority is as follows: `Global` < `VirtualHost` < `Route`
```

Comment:
It is not completely clear to me when processors are suggested to be combined. User must grasp the processors defined on different levels of configuration, how overrides and priority impacts their execution. The syntax also becomes quite verbose when external processing is configured on a route. 

I wonder if there would be simpler approach?

For example, in case of JWT verification the route-level verification policy refers to policies which are all defined at virtualhost level ([link](https://projectcontour.io/docs/1.27/config/jwt-verification/)). Also in case of external authorization, the extension service itself is not repeated on route level, even though [some finer grained](https://projectcontour.io/docs/1.27/config/client-authorization/#scoping-authorization-policy-settings) contols are provided there.

Reply:
I designed `extProc` based on these two implementations as above. I divided its functions into three levels: `global`, `virtualhost`, and `route`. The priorities of these levels increase successively. The latter two levels also provide options to disable the corresponding functions. This design makes the functions complete, but also complex.

To be honest, I am not particularly clear how users will use them in combination (that's why I came up with this design). But based on our company's product requirements, this function is mainly needed at the `route level`. So, to make it easier to understand and use, we can just implement the `route-level` function in the first phase.  @skriss @sunjayBhatia @tsaarni   WDYT?

Reply:
Sorry for the delay @izturn! 😳

In case of external authorization there is one reference to `ExtensionService` from top level  `HTTPProxy.spec.virtualhost.authorization`. On the route level `HTTPProxy.spec.routes.authPolicy` one can opt out, and add key:value pairs that will provide [context for the external authorization server](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#envoy-v3-api-field-extensions-filters-http-ext-authz-v3-checksettings-context-extensions) about the requests.

If copying the same idea here, could it be like this:

* Extension service would be configured only once under `HTTPProxy.spec.virtualhost.extProc` and it automatically applies to all routes
* Under `HTTPProxy.spec.routes.extProcPolicy` overrides could have only minor changes, such as disabling external processing for specific route but not e.g. replace the gRPC service with another one.

As far as I can see, the overrides in external authz are [just these](https://github.com/projectcontour/contour/blob/080940778e4995aa310e8d5e9caab5a0f2747c08/internal/envoy/v3/route.go#L184-L207).

Reply:
After our internal discussion and exploration, we found that the global ext_proc may not have much significance at the moment, so I have updated  the design. PTAL @tsaarni

Reply:
hey @tsaarni   Based on the positioning of ext_proc as an escape hatch, as well as discussions and explorations within our team, I've simplified the design: only support up to one ext_proc at each level, removing the design of adding multiple ext_proc(s) at different phases of the filter by priority, PTAL

Reply:
Currently route level override allows overriding everything. For example, it is possible to point each route to be processed by different external processor gRPC server. I'm curious, do you need this flexibility?

Reply:
https://github.com/projectcontour/contour/pull/5866#issuecomment-2021856492 @tsaarni

Reply:
+1 to what @tsaarni says, do we want to define also a separate servers? having the route setting override makes sense

>For some real use cases I meet, I think API/Route level customize logic is useful, because in some tranditional company, some of thier applications are not follow the standard of microservices and may have strange logics, so they need the Route level customize process to help only one API to do something, and rest of APIs are all work fine without this process, so I think if we can support Route level ext process, it will better.

would this not be achievable with just the overrides + 1 globally defined ext proc server?